HW2

hirokuu opened the pull request:
- refactored code according to the SOLID, KISS, DRY principles. used DI where is requires
- implement Swift API guidelines

hirokuu committed:
- refactored code according to the SOLID, KISS, DRY principles. use DI where is requires
- implement Swift API guidelines

hirokuu committed:
uncomment pod in the gitignore and fix method naming

AnnaSahaidak commented on 8 Ball/BD/PredictionHistoryManager.swift:
```diff
@@ -10,7 +10,7 @@ import CoreData
 
 
 class PredictionHistoryManager{
```
```suggestion
class PredictionHistoryManager {
```

AnnaSahaidak commented on 8 Ball/BD/AnswerManager.swift:
```diff
@@ -29,6 +27,7 @@ class AnswerManager {
     }
+    
     private static func isAnswerExist(_ answerString: String, type: AnswerType, createdByUser: Bool, context: NSManagedObjectContext) -> Bool {
         
```
Лишний перенос строки

hirokuu replied:
исправил

AnnaSahaidak commented on 8 Ball/Models/AnswerType.swift:
```diff
@@ -0,0 +1,53 @@
+        }
+    }
+    
+    init(fromString: String){
```
```suggestion
    init(fromString: String) {
```

AnnaSahaidak commented on 8 Ball/Models/AnswerType.swift:
```diff
@@ -0,0 +1,53 @@
+    case affirmative = 3
+    
+    
+    init(rawValue value: Int16) {
```
Этот инит дублирует дефолтный, не вижу смысла его делать. Чтобы задать дефолтное значение .unknown при создании нужно сделать так AnswerType(rawValue: someIntValue) ?? .unknown.

AnnaSahaidak commented on 8 Ball/Models/PredictionProviders/DBPredictionService.swift:
```diff
@@ -0,0 +1,27 @@
+    
+    func getPredictionForQuestion(_ question : String? = nil, completion: @escaping ((Predictible?, Error?) -> Void)) {
+        if let randomAnswer = AnswerManager.getRandomAnswer(context: context) {
+                completion(Prediction(question: question, answer: randomAnswer), nil)
```
Отступ съехал

AnnaSahaidak commented on 8 Ball/Models/PredictionProviders/PredictionProvider.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+protocol PredictionProvider {
```
👍🏼

AnnaSahaidak commented on 8 Ball/Models/PredictionProviders/RemotePredictionService.swift:
```diff
@@ -0,0 +1,46 @@
+
+class RemotePredictionService: PredictionProvider {
+    
+    private let networkDataProvider: NetworkDataProvider!
```
Можно не implicitly unwrapped делать, объявление в ините, не нужен `!`

AnnaSahaidak commented on 8 Ball/Models/PredictionProviders/RemotePredictionService.swift:
```diff
@@ -0,0 +1,46 @@
+    }
+    
+    func getPredictionForQuestion(_ question : String? = nil, completion: @escaping ((Predictible?, Error?) -> Void)) {
+        
```
Лишняя строка

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -7,44 +7,42 @@
 import Foundation
 
-enum PredictionMode{
+enum PredictionServiceMode{
```
```suggestion
enum PredictionServiceMode {
```

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -7,44 +7,42 @@
+    private let coreDataManager = CoreDataManager.shared
+    private var predictionProvider: PredictionProvider?
+    
+    private var delegate:PredictionServiceDelegate{
```
```suggestion
    private var delegate: PredictionServiceDelegate {
```

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -7,44 +7,42 @@
+    init(delegate: PredictionServiceDelegate) {
+        self.delegate = delegate
+        delegate.setPredictionMode(mode)
+        self.predictionProvider = RemotePredictionService(networkDataProvider: NetworkService())
```
Эту зависимость стоит инжектить

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -99,90 +103,49 @@ class PredictionService {
-    
-    private func servicePrediction() throws {
-        EightBallPredictionService.getPredictionForQuestion(question, completion: {(magicData, error) in
+        predictionProvider?.getPredictionForQuestion(question, completion: {(prediction, error) in
```
```suggestion
        predictionProvider?.getPredictionForQuestion(question, completion: { (prediction, error) in
```

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -99,90 +103,49 @@ class PredictionService {
-    private func servicePrediction() throws {
-        EightBallPredictionService.getPredictionForQuestion(question, completion: {(magicData, error) in
+        predictionProvider?.getPredictionForQuestion(question, completion: {(prediction, error) in
             
```
Лишняя строка

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -99,90 +103,49 @@ class PredictionService {
-            if let magicData = magicData {
+            
+            if let prediction = prediction {
                 
```
Лишняя строка

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -99,90 +103,49 @@ class PredictionService {
                 if self.mode == .question {
                     //Add question with answer to the DB
                     do {
                         
```
Лишняя строка

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -99,90 +103,49 @@ class PredictionService {
-            throw PredictionServiceError.customAnswerNotFound
-        }
+    public func showPrediction() {
+        self.isReadyToShow = true
```
Писать self  не обязательно, ниже в методе тоже

AnnaSahaidak commented on 8 Ball/Modules/AnswerEditor/AnswerEditorView.swift:
```diff
@@ -1,35 +1,44 @@
+class AnswerEditorView: UIView {
     
-    var delegate : AnswerListRefreshProtocol!
+   private var answerEditorViewControllerInput: AnswerEditorViewControllerInput!
```
Не обязательно делать implicitly unwrapped, объявляется в ините, можно не писать `!` и сделать let

AnnaSahaidak commented on 8 Ball/Modules/AnswerEditor/AnswerEditorView.swift:
```diff
@@ -100,130 +108,98 @@ class AnswerEditorViewController: UIViewController, UITextFieldDelegate {
+    init(frame: CGRect, delegate: AnswerEditorViewControllerInput) {
+        super.init(frame: frame)
+        self.answerEditorViewControllerInput = delegate
+        self.addCustomView()
```
Можно не писать self
И отдели пустой строкой super вызов, это визуально отделяет код.

AnnaSahaidak commented on 8 Ball/Modules/AnswerEditor/AnswerEditorViewController.swift:
```diff
@@ -0,0 +1,82 @@
+    
+    var answer: Answer?
+    
+    private var answerListListener : AnswerListListener!
```
```suggestion
    private let answerListListener : AnswerListListener
```

AnnaSahaidak commented on 8 Ball/Modules/AnswerEditor/AnswerEditorViewController.swift:
```diff
@@ -0,0 +1,82 @@
+        view.addSubview(AnswerEditorView(frame: view.frame, delegate: self))
+        answerEditorViewInput.setTextFieldDelegate(self)
+        
+        guard let answer = answer else {return}
```
```suggestion
        guard let answer = answer else { return }
```

AnnaSahaidak commented on 8 Ball/Modules/AnswerList/AnswerListRouter.swift:
```diff
@@ -0,0 +1,39 @@
+class AnswerListRouter: Router {
+    typealias RoutingDestination = AnswerListDestination
+    
+    private unowned var viewController: UIViewController!
```
```suggestion
    private let viewController: UIViewController
```

hirokuu replied:
если не сделать сcылку слабой, то получиться замыкание сильных ссылок  router->view controller, viewController->router

```suggestion
    private unowned let viewController: UIViewController
```

AnnaSahaidak replied:
Да, ты прав, не заметила что друг на друга ссылаются

AnnaSahaidak commented on 8 Ball/Modules/Prediction/PredictionViewController.swift:
```diff
@@ -0,0 +1,133 @@
+
+class PredictionViewController: UIViewController {
+    
+    var model: PredictionService!
```
Может переменную service назвать, а то как-то странно model тут выглядит)?
```suggestion
    private let model: PredictionService
```

AnnaSahaidak commented on 8 Ball/Modules/Prediction/PredictionViewController.swift:
```diff
@@ -0,0 +1,133 @@
+        questionLabel.text = ""
+        typeLabel.text = ""
+        
+        model = PredictionService(delegate: self)
```
Лучше эту зависимость инжектить, пусть и через проверти, но точно не тут объявлять

AnnaSahaidak commented on 8 Ball/Modules/Prediction/PredictionViewController.swift:
```diff
@@ -0,0 +1,133 @@
+        typeLabel.text = ""
+        
+        model = PredictionService(delegate: self)
+        router = PredictionRouter(viewController: self)
```
Лучше эту зависимость инжектить, пусть и через проверти, но точно не тут объявлять

AnnaSahaidak replied:
Обе переменные (service/router) нужно инжектить

AnnaSahaidak commented on 8 Ball/Modules/Settings/SettingsTableViewController.swift:
```diff
@@ -13,72 +13,68 @@ enum SettingsList: String{
 }
 
-protocol SettingsSetProtocol{
+protocol PredictionSettingsConfiguration{
```
```suggestion
protocol PredictionSettingsConfiguration {
```

AnnaSahaidak commented on 8 Ball/Modules/Settings/SettingsTableViewController.swift:
```diff
@@ -13,72 +13,68 @@ enum SettingsList: String{
     
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        router = SettingsRouter(viewController: self)
```
Нужно инжектить

AnnaSahaidak commented on 8 Ball/Modules/Settings/SettingsTableViewController.swift:
```diff
@@ -13,72 +13,68 @@ enum SettingsList: String{
 
     // MARK: - Table view data source
-
     override func numberOfSections(in tableView: UITableView) -> Int {
```
Ты правильно выносил реализацию протоколов в extensions, вынеси и методы делегата и датасоурса таблицы тоже

AnnaSahaidak replied:
Это можно сделать в рамках следующего дз (вынос методов в extension)

AnnaSahaidak commented on 8 Ball/Networking/NetworkService.swift:
```diff
@@ -0,0 +1,27 @@
+import Foundation
+import Alamofire
+
+protocol NetworkDataProvider {
```
👍🏼

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -7,44 +7,42 @@
     
-    private let cdm = CoreDataManager.shared
+    private let coreDataManager = CoreDataManager.shared
+    private var predictionProvider: PredictionProvider?
```
Можно объявить как let и не опционалом

AnnaSahaidak commented on 8 Ball/Modules/AnswerEditor/AnswerEditorView.swift:
```diff
@@ -100,130 +108,98 @@ class AnswerEditorViewController: UIViewController, UITextFieldDelegate {
+        discardButton.addTarget(self, action:  #selector(self.dismissAction), for: .touchUpInside)
         
-        configureUIForAnswer()
+        applyButton.addTarget(self, action:  #selector(self.applyAction), for: .touchUpInside)
```
```suggestion
        applyButton.addTarget(self, action: #selector(applyAction), for: .touchUpInside)
```

AnnaSahaidak commented on 8 Ball/Modules/AnswerEditor/AnswerEditorViewController.swift:
```diff
@@ -0,0 +1,82 @@
+    
+    private var answerListListener : AnswerListListener!
+    
+    lazy var answerEditorViewInput: AnswerEditorViewInput  = { return view.subviews.first! as! AnswerEditorViewInput }()
```
Можно это вынести в аутлеты? Не нравится два форса, стоит избегать такого.

AnnaSahaidak commented on 8 Ball/Modules/AnswerList/AnswerListTableViewController.swift:
```diff
@@ -9,28 +9,27 @@ import UIKit
-    let context = CoreDataManager.shared.persistentContainer.viewContext
+class AnswerListTableViewController: UITableViewController {
+    
+    private let coreDataManager = CoreDataManager.shared
```
На прошлой лекции как раз рассказывали что это не ответственность вью контроллера обращаться напрямую к базе данных, это работа для сервиса.

hirokuu replied:
теперь этот функционал выполняет AnswerManager

AnnaSahaidak replied:
Значит переменные больше не нужны? Удали их тогда.

AnnaSahaidak commented on 8 Ball/Modules/AnswerList/AnswerListTableViewController.swift:
```diff
@@ -9,28 +9,27 @@ import UIKit
+    private let coreDataManager = CoreDataManager.shared
+    private let context = CoreDataManager.shared.persistentContainer.viewContext
+    
+    private var router : AnswerListRouter!
```
```suggestion
    private let router: AnswerListRouter
```

AnnaSahaidak commented on 8 Ball/Modules/AnswerList/AnswerListTableViewController.swift:
```diff
@@ -9,28 +9,27 @@ import UIKit
     override func viewDidLoad() {
         super.viewDidLoad()
-        
+        router = AnswerListRouter(viewController: self)
```
Нужно инжектить

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -56,12 +54,18 @@ class PredictionService {
     
     @objc private func setUseCustomAnswers() {
         useCustomAnswers = true
+        predictionProvider = DBPredictionService(context: coreDataManager.persistentContainer.viewContext)
```
Плохое решение внутри подменять провайдера. По хорошему в одном случае должен создаваться сервис с одним провайдером, в другом случае с другим. Или же иметь два провайдера и в зависимости от каких-то настроек получать ответы от одного из них.

AnnaSahaidak commented on 8 Ball/Modules/Settings/SettingsTableViewController.swift:
```diff
@@ -13,72 +13,68 @@ enum SettingsList: String{
 class SettingsTableViewController: UITableViewController {
 
     private let data : [SettingsList] = [.useCustomAnswers, .answers, .predictionHistory]
+    private var router: SettingsRouter?
```
```suggestion
    private let router: SettingsRouter
```

AnnaSahaidak reviewed (changes requested):
В целом, проделана хорошая работа.
Есть незначительные комментарии к стилистике кода, важно следить за выравниваниями кода, отступами. 
Недоработаны во всех местах dependency injection.
Так же присутствует обращение к CoreData из контроллеров, от этого нужно уйти, fetch requests вынести в какой-то сервис (уже есть сервис DBPredictionService он по идее и спроектирован работать с БД)
Еще заметила использование модификатора public, на сегоднешней лекции думаю стало понятно что он не нужен в этом проекте)
И еще реализацию протокола в extension не нужно дублировать MARK и так все видно из протокола, не стоит описывать очевидные вещи.

AnnaSahaidak commented on 8 Ball/Cells/SettingsTableViewCell.swift:
```diff
@@ -10,33 +10,28 @@ import UIKit
 class SettingsTableViewCell: UITableViewCell {
     
-//    var dataItem: SettingsList!
+    private var delegate: PredictionSettingsConfiguration?
```
```suggestion
    private let delegate: PredictionSettingsConfiguration
```

AnnaSahaidak commented on 8 Ball/Modules/Prediction/PredictionRouter.swift:
```diff
@@ -0,0 +1,54 @@
+class PredictionRouter: Router {
+    typealias RoutingDestination = PredictionDestination
+    
+    private unowned var viewController: UIViewController!
```
```suggestion
    private ley viewController: UIViewController
```

hirokuu committed:
- removed empty lined and extra white spaces
- replaced public access modificatory to private where its possible
- removed init(rawValue value: Int16) from AnswerType
- replaced var to let for properties that was initialized in init

hirokuu committed:
PredictionService:
- implemented DI to predictionProvider
- implemented methods to passing data from/to PredictionService through the NotificationCenter

hirokuu committed:
added custom init instead of function with DI in SettingsTableViewCell

hirokuu committed:
moved DB interaction from AnswerEditorViewController to AnswerManager

AnnaSahaidak commented on 8 Ball/BD/AnswerManager.swift:
```diff
@@ -8,11 +8,62 @@
+    private let delegate: AnswerListPresenter
+    private let context: NSManagedObjectContext
+    
+    private lazy var fetchedResultsController : NSFetchedResultsController<Answer> = {
```
```suggestion
    private lazy var fetchedResultsController: NSFetchedResultsController<Answer> = {
```

AnnaSahaidak commented on 8 Ball/BD/AnswerManager.swift:
```diff
@@ -8,11 +8,62 @@
+    private let context: NSManagedObjectContext
+    
+    private lazy var fetchedResultsController : NSFetchedResultsController<Answer> = {
+        let fetchRequest : NSFetchRequest<Answer> = NSFetchRequest<Answer>(entityName: Answer.entity().name!)
```
```suggestion
        let fetchRequest: NSFetchRequest<Answer> = NSFetchRequest<Answer>(entityName: Answer.entity().name!)
```

AnnaSahaidak commented on 8 Ball/BD/AnswerManager.swift:
```diff
@@ -8,11 +8,62 @@
+}
+
+extension AnswerManager {
     static func getOrCreateAnswer(_ answerString: String, type: AnswerType, createdByUser: Bool, context: NSManagedObjectContext) -> Answer {
```
Плохо что методы в extension static, в задании говорилось уйти по возможности от всех статических функций

hirokuu replied:
не совсем понимаю почему создание или редактирование экземпляра класса требует создания экземпляра класса менеджера. если через статические методы это вполне удобно делать)

AnnaSahaidak replied:
Зачастую правильные с точки зрения архитектуры и масштабирования решения как раз менее удобны и требуют больших усилий, но это только на первый взгляд, в дальнейшем ощущается огромная польза от хорошо спроектированного кода. 
В целом, стоит минимизировать использование синглтонов и статических методов, это даже стало антипаттерном. Смотри, статический метод ты можешь вызвать откуда угодно, таким образом легко нарушить принципы SOLID, нарушив ответственность и вызвав статические методы откуда не должны этого делать. Опять же используя статические методы ты не инжектишь зависимость, получаем сильно связанный код, со всеми вытекающими отсюда последствиями.

AnnaSahaidak replied:
Если бы это была какая-то фабрика создания комплексных объектов (factory pattern) - статические методы создания объектов были бы уместны, но это не тот случай)

AnnaSahaidak commented on 8 Ball/BD/PredictionHistoryManager.swift:
```diff
@@ -9,8 +9,8 @@ import Foundation
-class PredictionHistoryManager{
-    static func createPediction(question: String, answer: Answer, context: NSManagedObjectContext) {
+class PredictionHistoryManager {
+    static func createPrediction(question: String, answer: Answer, context: NSManagedObjectContext) {
```
Тут тоже, обязательно ли делать статической функцию?

AnnaSahaidak commented on 8 Ball/Modules/PredictionHistoryListTableViewController.swift:
```diff
@@ -8,21 +8,18 @@
+    let coreDataManager = CoreDataManager.shared
     let context = CoreDataManager.shared.persistentContainer.viewContext
     
     lazy var fetchedResultsController : NSFetchedResultsController<PredictionHistory> = {
```
А тут обращение к CoreData не переделал на обращение к сервису

AnnaSahaidak commented on 8 Ball/Modules/PredictionHistoryListTableViewController.swift:
```diff
@@ -62,25 +69,21 @@ class PredictionsHistoryTableViewController: UITableViewController {
             do {
                 try PredictionHistoryManager.deletePrediction(fetchedResultsController.object(at: indexPath), context: context)
-                try cdm.saveContext(context)
+                try coreDataManager.saveContext(context)
```
Снова обращение к БД из view controllera

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -7,61 +7,74 @@
-    @objc private func setUseCustomAnswers() {
-        useCustomAnswers = true
+    @objc private func getIsUseCustomAnswers() {
+        NotificationCenter.default.post(name: .setUseCustomAnswersTo, object: nil, userInfo: ["setUseCustomAnswersTo": useCustomAnswers])
```
Использование нотификаций заманчиво, но это плохая практика так их использовать. Дело в том что подписаться на такую нотификацию может какой угодно класс/структура и отдебажить такое потом сложно, не очевидным становится где это все обрабатывается, да и легко нарушить ответственности и обрабатывать нотификации не там где это стоит делать. Тем более класс, который отправляет нотификацию сам же и обрабатывает ее, очень странно), зачем так запутывать, ведь можно просто через вызов метода все сделать. 
Переделка на твое усмотрение, просто говорю что так применять нотификации - плохая практика.

hirokuu replied:
сделал нотификации из-за того что метод который переключает predictionProvider вызывается из вью контроллера который находится на другой Табе. Можно конечно это исправить используя синглтон(такой вариант прозрачнее), но в задании просили от них избавится

AnnaSahaidak replied:
Я бы хотела тут видеть другое решение, без синглтона. Давай так, будет лекция про архитектруры, завтра, там в рамках дз нужно будет внести изменения, я думаю после этого у тебя появятся идеи как это можно изменить. Отложим этот момент пока, а потом если что я опишу как я вижу это.

AnnaSahaidak commented on 8 Ball/Modules/AnswerEditor/AnswerEditorView.swift:
```diff
@@ -100,130 +108,100 @@ class AnswerEditorViewController: UIViewController, UITextFieldDelegate {
+        
+        answerTextField.text = answer.title
+        
+        if answer.type >= 0 && answer.type <= 3 {
```
От magic numbers стоит отказываться, тем более у тебя есть enum для типа ответа. (исправление опционально, просто рекомендация)

AnnaSahaidak commented on 8 Ball/Modules/AnswerEditor/AnswerEditorView.swift:
```diff
@@ -100,130 +108,100 @@ class AnswerEditorViewController: UIViewController, UITextFieldDelegate {
+//MARK: - AnswerEditorViewOutput
+extension AnswerEditorView: AnswerEditorViewOutput {
+    func getAnswer() -> String {
+        return answerTextField.text!
```
Лучше дать дефолтное значение "", чем force unwrap, старайся вообще не использовать force unwrap (можно исправить в следующем дз)

AnnaSahaidak commented on 8 Ball/Modules/AnswerEditor/AnswerEditorView.swift:
```diff
@@ -100,130 +108,100 @@ class AnswerEditorViewController: UIViewController, UITextFieldDelegate {
-        textField.resignFirstResponder()
-        return true
+    func getType() -> AnswerType {
+        return AnswerType(rawValue: Int16(typeSegmentedControl.selectedSegmentIndex)) ?? .unknown
```
Вот, хотела еще сразу написать про тип enum `AnswerType`, но не стала. Можно было использовать тип `Int` и избежать вот таких приведений типов, на практике используется просто `Int`, только в каких-то специфических случаях, где это действительно важно стоит использовать более конкретный тип.

AnnaSahaidak reviewed (changes requested):
Не все комментарии исправлены, не все зависимости инжектятся, не везде вынесена работа с БД из контроллеров

hirokuu committed:
- changed AnswerType.rawValue type from Int16 to Int
- removed useless properties

hirokuu committed:
fix bug when user was not able to change answer type

hirokuu committed:
implement all Answer related methods in AnswerListManaer class

hirokuu committed:
implement all PredictionHistory related methods in PredictionHistoryManager class

AnnaSahaidak commented on 8 Ball/Models/AnswerType.swift:
```diff
@@ -0,0 +1,42 @@
+import Foundation
+
+enum AnswerType: Int, CaseIterable {
+    case unknown = 0
```
И еще, просто чтоб ты знал, по дефолту у enum типа Int значения и так будут присвоены начиная с нуля, с итерацией в 1, т.е. точно так как ты указал. Явно указывают значение если начало нумерации не с нуля или коды вообще какие-то специфически (например, какие-то коды ошибок к бэка 1001 и т.п.)

hirokuu replied:
занимательно)

AnnaSahaidak commented on 8 Ball/Models/PredictionService.swift:
```diff
@@ -85,104 +98,58 @@ class PredictionService {
-        EightBallPredictionService.getPredictionForQuestion(question, completion: {(magicData, error) in
-            
+        predictionProvider.getPredictionForQuestion(question, completion: { [weak self] (prediction, error) in
+            guard let `self` = self else {return}
```
В этой версии swift уже не нужно писать кавычки, просто self = self

AnnaSahaidak commented on 8 Ball/Modules/AnswerEditor/AnswerEditorViewController.swift:
```diff
@@ -0,0 +1,83 @@
+                
+                let type: AnswerType = answerEditorViewOutput.getType()
+                
+                let context = CoreDataManager.shared.persistentContainer.viewContext
```
Использование не вижу, видимо просто не удалил

hirokuu committed:
implemented EntityListProvider<Entity> to avoid code repeating for Answer and PredictionHistory lists
